### PR TITLE
add support for Layout Management

### LINKED ISSUE #7: Update to add support for Layout Management
WakeUp looks tiny on modern screens, and nearly all elements are cut off. The window is also not resizable.

It does not utilize the Layout API, therefore, it should be updated.

image

### PR DESCRIPTION
**Pull request by Ra-V2**
Fixes #7

### CODE REVIEW

**Comment by owenca** [comment]
Setting fonts to 8 points doesn't look good:
layout

**Comment by owenca** [comment]
Also, make the input box narrower and group it with "min" so that the gap between them is not too wide.

File: buzzwindow.cpp [hunk 1]
```diff
@@ -5,33 +5,51 @@
...
-	StartButton = new BButton(BRect(1, 25, 35, 25), "", "Start", new BMessage(START));
-	AddChild(StartButton);
+	StartButton = new BButton("", "Start", new BMessage(START));
 
-	IntervalControl = new BTextControl(BRect(35, 30, 90, 40), "", "", "5", 
-										NULL, B_FOLLOW_NONE, B_WILL_DRAW);
-	IntervalControl->SetDivider(0);
-	AddChild(IntervalControl);
+	IntervalControl = new BTextControl("", "", "5", 
+										NULL, B_WILL_DRAW);
```
Comment (owenca) [inline buzzwindow.cpp]: Merge them into one line.

**Review by owenca — CHANGES_REQUESTED** [review]
When the user press the Start button, the timer is supposed to be displayed above the button, erasing the Test button and the sound file name. Hence you have to swap them dynamically.

File: buzzwindow.cpp [hunk 2]
```diff
@@ -5,33 +5,49 @@
...
 #include <Application.h>
 #include "main.h"
+#include <GroupLayout.h>
+#include <GroupLayoutBuilder.h>
+#include <LayoutBuilder.h>
+
 //-------------------------------------------------------------------
 BuzzWindow::BuzzWindow()
-	: BWindow(BRect(30, 30, 150, 80), "WakeUp v1.0", B_TITLED_WINDOW, B_NOT_RESIZABLE | B_NOT_ZOOMABLE)
+	: BWindow(BRect(30, 30, 0, 0), "WakeUp v1.0", B_TITLED_WINDOW, B_NOT_RESIZABLE | B_NOT_ZOOMABLE | B_AUTO_UPDATE_SIZE_LIMITS)
```
Comment (owenca) [inline buzzwindow.cpp]: Coding style.

**Review by owenca — CHANGES_REQUESTED** [review]
There are two problems now:
- The window shrinks in height when the timer is displayed.
- The widgets are not centered anymore.

File: buzzwindow.cpp [hunk 3]
```diff
@@ -5,33 +5,45 @@
 #include "enum.h"
 #include <Application.h>
 #include "main.h"
+#include <GroupLayout.h>
+#include <GroupLayoutBuilder.h>
```
Comment (owenca) [inline buzzwindow.cpp]: You don't need these two lines.

File: buzzwindow.cpp [hunk 4]
```diff
@@ -5,33 +5,45 @@
...
+#include <LayoutBuilder.h>
+
 //-------------------------------------------------------------------
 BuzzWindow::BuzzWindow()
-	: BWindow(BRect(30, 30, 150, 80), "WakeUp v1.0", B_TITLED_WINDOW, B_NOT_RESIZABLE | B_NOT_ZOOMABLE)
+	: BWindow(BRect(30, 30, 0, 0), "WakeUp v1.0",
```
Comment (owenca) [inline buzzwindow.cpp]: The `:` should be on its own line. Please refer to [Haiku Coding Guidelines](https://www.haiku-os.org/development/coding-guidelines/).

File: buzzwindow.cpp [hunk 5]
```diff
@@ -5,33 +5,45 @@
...
 
-	Seconds = new BStringView(BRect(90,35,130,45), "", "min");
-	AddChild(Seconds);
+	Seconds = new BStringView("", "min");
 
-	TimeElapsed = new BStringView(BRect(1,5,130,25), "", "0:0:0");
+	TimeElapsed = new BStringView("", "0:0:0");
 	TimeElapsed->SetFont(be_bold_font);
-	TimeElapsed->SetFontSize(20);
+	TimeElapsed->SetFontSize(22);
```
Comment (owenca) [inline buzzwindow.cpp:28]: This only works with the 12-point font size. We will live with it for now, but can you please create a new issue for this?

File: buzzwindow.cpp [hunk 6]
```diff
@@ -5,33 +5,45 @@
...
+	BLayoutBuilder::Group<>(this, B_VERTICAL, B_USE_ITEM_SPACING)
+		.SetInsets(B_USE_WINDOW_INSETS)
+		.AddGroup(B_HORIZONTAL)
+			.Add(TimeElapsed)
+			.Add(TestButton)
+			.Add(SoundName)
+			.End()
+		.AddGroup(B_HORIZONTAL)
+			.Add(StartButton)
+			.AddGroup(B_HORIZONTAL, 1)
```
Comment (owenca) [inline buzzwindow.cpp]: Can you use an appropriate `B_...` macro for the spacing value instead?

**Review by owenca — CHANGES_REQUESTED** [review]

**Comment by Ra-V2** [comment]
I added AddGlue() because when we drop a file with small amount of characters name, the test button and file name get centered and this not looks good for me.

File: buzzwindow.cpp [hunk 7]
```diff
@@ -5,33 +5,45 @@
 #include "enum.h"
 #include <Application.h>
 #include "main.h"
+#include <LayoutBuilder.h>
 //-------------------------------------------------------------------
 BuzzWindow::BuzzWindow()
-	: BWindow(BRect(30, 30, 150, 80), "WakeUp v1.0", B_TITLED_WINDOW, B_NOT_RESIZABLE | B_NOT_ZOOMABLE)
+	:
+	BWindow(BRect(30, 30, 0, 0), "WakeUp v1.0",
+	B_TITLED_WINDOW, B_NOT_RESIZABLE | B_NOT_ZOOMABLE | B_AUTO_UPDATE_SIZE_LIMITS)
```
Comment (owenca) [inline buzzwindow.cpp]:
```
	BWindow(BRect(30, 30, 0, 0), "WakeUp v1.0", B_TITLED_WINDOW,
		B_NOT_RESIZABLE | B_NOT_ZOOMABLE | B_AUTO_UPDATE_SIZE_LIMITS)
```

File: buzzwindow.cpp [hunk 8]
```diff
@@ -5,33 +5,45 @@
...
 
-	StartButton = new BButton(BRect(1, 25, 35, 25), "", "Start", new BMessage(START));
-	AddChild(StartButton);
+	StartButton = new BButton("", "Start", new BMessage(START));
 
-	IntervalControl = new BTextControl(BRect(35, 30, 90, 40), "", "", "5", 
-										NULL, B_FOLLOW_NONE, B_WILL_DRAW);
-	IntervalControl->SetDivider(0);
-	AddChild(IntervalControl);
+	IntervalControl = new BTextControl("", "", "5", NULL, B_WILL_DRAW);
```
Comment (owenca) [inline buzzwindow.cpp:21]: IMO, the time value should be right aligned.

File: buzzwindow.cpp [hunk 9]
```diff
@@ -5,33 +5,45 @@
...
+		.AddGroup(B_HORIZONTAL)
+			.Add(TestButton)
+			.Add(SoundName)
+			.AddGlue()
+			.Add(TimeElapsed)
+			.AddGlue()
+			.End()
+		.AddGroup(B_HORIZONTAL)
+			.Add(StartButton)
+			.AddGroup(B_HORIZONTAL, B_TRACK_RECT_CORNER)
```
Comment (owenca) [inline buzzwindow.cpp]: I suggest B_USE_HALF_ITEM_SPACING here.

**Review by owenca — CHANGES_REQUESTED** [review]

**Review by owenca — CHANGES_REQUESTED** [review]
Can you update WakeUp.png and squash the commits? We can merge the PR after that.

Edit: Also make sure that ReadMe.md will show the new screenshot.

**Review by owenca — APPROVED** [review]

**Pull request merged**